pkg/brbdxr/brbdxr.go

```diff
@@ -0,0 +1,289 @@
+
+				data := []byte{0, 0, 0, 0}
+				binary.LittleEndian.PutUint32(data, uint32(len(hdata)))
+				data = append(data, hdata...)
+
+				dsl.HashOneMessage(m, mc.Hasher, [][]byte{hdata}, &hashInitialMessageContext{id: id, data: data})
+				return nil
+			} else {
+				return fmt.Errorf("already sent echo")
+			}
+
+		} else {
+			return fmt.Errorf("received start message not from leader")
+		}
+	})
+
+	dsl.UponHashResult(m, func(hashes [][]byte, context *hashInitialMessageContext) error {
+		if context.id <= lastId {
+			return nil
+		}
+		chunkSize := int(math.Ceil(float64(len(context.data)) / float64(params.GetN()-2*params.GetF())))
+		encoded := make([][]byte, params.GetN())
+		for i := range encoded {
+			encoded[i] = make([]byte, chunkSize)
+		}
+
+		for i, in := range encoded[:params.GetN()-2*params.GetF()] {
+			for j := range in {
+				if i*chunkSize+j < len(context.data) {
+					in[j] = context.data[i*chunkSize+j]
+				} else {
+					in[j] = 0
+				}
+			}
+		}
+
+		err := encoder.Encode(encoded)
+
+		if err != nil {
+			return err
+		}
+
+		for i, node := range params.AllNodes {
+			eventpbdsl.SendMessage(m, mc.Net, brbdxrpbmsgs.EchoMessage(mc.Self, context.id, hashes[0], encoded[i]), []t.NodeID{node})
+		}
+		return nil
+	})
+
+	brbdxrpbdsl.UponEchoMessageReceived(m, func(from t.NodeID, id int64, hash, chunk []byte) error {
+		if id <= lastId {
+			return nil
+		}
+		initialize(&state, id, params.GetN())
+		fromId, err := strconv.ParseInt(from.Pb(), 10, 64)
+		if err != nil {
+			return err
+		}
+		if state[id].echos[fromId] == nil {
+			state[id].echos[fromId] = chunk
+
+			incrementAndUpdateAccumulator(&hash, &chunk, &state[id].echoMessagesCount, &state[id].echosMaxAccumulator)
+		}
+		return nil
+	})
+
+	brbdxrpbdsl.UponReadyMessageReceived(m, func(from t.NodeID, id int64, hash, chunk []byte) error {
+		if id <= lastId {
+			return nil
+		}
+		initialize(&state, id, params.GetN())
+		fromId, err := strconv.ParseInt(from.Pb(), 10, 64)
+		if err != nil {
+			return err
+		}
+		if state[id].readys[fromId] == nil {
+			state[id].readys[fromId] = chunk
+			state[id].readyMessagesReceived += 1
+
+			incrementAndUpdateAccumulator(&hash, &chunk, &state[id].readyMessagesCount, &state[id].readyMaxAccumulator)
+		}
+		return nil
+	})
+
+	dsl.UponCondition(m, func() error {
+		for id, currentState := range state {
+			if id <= lastId {
+				continue
+			}
+			if (currentState.echosMaxAccumulator.count > (params.GetN()+params.GetF())/2 ||
+				currentState.readyMaxAccumulator.count > params.GetF()) && currentState.sentReady == false {
+
+				currentState.sentReady = true
+				eventpbdsl.SendMessage(m, mc.Net, brbdxrpbmsgs.ReadyMessage(
+					mc.Self, id, *currentState.echosMaxAccumulator.hash,
+					*currentState.echosMaxAccumulator.chunk,
+				), params.AllNodes)
+			}
+
+			if currentState.readyMessagesReceived > 2*params.GetF() && currentState.delivered == false {
+
+				err := encoder.Reconstruct(currentState.readys)
```

Comment:
Here, it may be important to do the reconstruction in a separate module to avoid blocking the main protocol thread.